mmhuman3d/core/evaluation/eval_utils.py

```diff
@@ -197,3 +217,52 @@ def keypoint_3d_auc(pred, gt, mask, alignment='none'):
+    :param grundtruth_landmark_points: A 7 x 3 list with annotations of the ground truth scan.
+    :param predicted_mesh_vertices: An m x 3 numpy array of vertices from a predicted mesh.
+    :param predicted_mesh_faces: A k x 3 numpy array of vertex indices composing the predicted mesh.
+    :param predicted_mesh_landmark_points: A 7 x 3 list containing the annotated 3D point locations in the predicted mesh.
+    :return: A list of distances (errors), one for each vertex in the groundtruth mesh, and the associated vertex index in the ground truth scan.
+
+    The grundtruth_landmark_points and predicted_mesh_landmark_points have to contain points in the following order:
+    (1) right eye outer corner, (2) right eye inner corner, (3) left eye inner corner, (4) left eye outer corner,
+    (5) nose bottom, (6) right mouth corner, (7) left mouth corner.
+    """
+
+    # Do procrustes based on the 7 points:
+    # The ground truth scan is in mm, so by aligning the prediction to the ground truth, we get meaningful units.
+    _,tform = compute_similarity_transform(predicted_mesh_landmark_points, grundtruth_landmark_points, return_tform= True)
+    # Use tform to transform all vertices in predicted_mesh_vertices to the ground truth reference space:
+    predicted_mesh_vertices_aligned = (tform['scale'] * tform['rotation'].dot(predicted_mesh_vertices.T) + tform['translation']).T
+    
+    # Compute the mask: A circular area around the center of the face. Take the nose-bottom and go upwards a bit:
+    nose_bottom = np.array(grundtruth_landmark_points[4])
+    nose_bridge = (np.array(grundtruth_landmark_points[1]) + np.array(
+        grundtruth_landmark_points[2])) / 2  # between the inner eye corners
+    face_centre = nose_bottom + 0.3 * (nose_bridge - nose_bottom)
+    # Compute the radius for the face mask:
+    outer_eye_dist = np.linalg.norm(np.array(grundtruth_landmark_points[0]) - np.array(grundtruth_landmark_points[3]))
+    nose_dist = np.linalg.norm(nose_bridge - nose_bottom)
+    mask_radius = 1.2 * (outer_eye_dist + nose_dist) / 2
+
+    # Find all the vertex indices in the ground truth scan that lie within the mask area:
+    vertex_indices_mask = []  # vertex indices in the source mesh (the ground truth scan)
+    points_on_groundtruth_scan_to_measure_from = []
+    for vertex_idx, vertex in enumerate(groundtruth_vertices):
+        dist = np.linalg.norm(vertex - face_centre) # We use Euclidean distance for the mask area for now.
+        if dist <= mask_radius:
+            vertex_indices_mask.append(vertex_idx)
+            points_on_groundtruth_scan_to_measure_from.append(vertex)
+    assert len(vertex_indices_mask) == len(points_on_groundtruth_scan_to_measure_from)
+    # For each vertex on the ground truth mesh, find the closest point on the surface of the predicted mesh:
+    predicted_mesh_pymesh = pymesh.meshio.form_mesh(np.array(predicted_mesh_vertices_aligned), predicted_mesh_faces)
```

Comment:
Can we try using `trimesh` instead of `pymesh`?

Reply:
It is now changed to trimesh.